### ISSUE TITLE
In JupyterLab 4, RTC for a file inside custom drive fails save.

### ISSUE THREAD

**Issue by a3626a** [jupyterlab/jupyterlab#13203]
## Description
When a collaborative file is located in custom drive, opening the file leads following error (in jupyter server console)
```
[W 2022-10-09 14:25:59.226 ServerApp] 404 GET /api/yjs/json:notebook:{mydrive}:Untitled.ipynb (127.0.0.1): file or directory does not exist: '{mydrive}:Untitled.ipynb'
```

RTC is still available, but it does not save the file.

https://user-images.githubusercontent.com/4434752/194739976-76c9f820-5a1d-4b57-a35f-f8fe206be496.mp4

## Reproduce

1. Register Drive
```
// This is just default drive, so it access file from file system.
app.serviceManager.contents.addDrive(new Drive({ name: 'mydrive' }));
```
2. Open collaborative file using drive path
`http://localhost:8888/lab/tree/mydrive:Untitled.ipynb`

## Expected behavior

Error should not be raised. File must be saved.

## Context

- Operating System and version:
It is WSL.
```
lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 20.04 LTS
Release:        20.04
Codename:       focal
```
- Browser and version: Chrome 106.0.5249.103
- JupyterLab version: `v4.0.0a29`

## Analysis

`/api/yjs/json` API does not support drive. Acutally, 'drive' is an additional layer before the jupyter server, jupyter server should not be aware of the drive.
-> So, at least, drive should not be sent to server.

One easy way is just to make files inside custom drvie not collaborative.

To fix this issue completly, Drive interface should provide a new function to tell proper path, or to create a doc provider or doc provider factory.

the new function would be called here:
https://github.com/jupyterlab/jupyterlab/blob/25761d3c2d8dcb3611998e2fc1c9ddee85c187f2/packages/docregistry/src/context.ts#L67-L75

**Comment by jtpio** [comment]
Thanks @a3626a for reporting and providing helpful details :+1: 

> > So, at least, drive should not be sent to server.
> 
> One easy way is just to make files inside custom drvie not collaborative.

I think it makes sense to not trigger any request to the server. These custom drives might indeed be manipulating files from elsewhere, or even just read-only data like with https://github.com/jupyterlab/jupyterlab-github.

**Comment by davidbrochart** [comment]
This does question our current design though. In order for a document to be collaborative, a connection to `/api/yjs/...` has to be made (`YDocWebSocketHandler`), which also handles loading/saving the document. In other words, the back-end provides the document.
The contents manager in the back-end could talk to an external "drive" (e.g. a GitHub repo), but I understand that the front-end can directly talk to this drive through custom drives. In this case, the front-end provides the document.
I'm wondering if we should handle "external shared documents" in the back-end, so that front-ends can still use all the back-end machinery for propagating updates to all users. Not to mention authorization which is really a back-end thing.

**Comment by jtpio** [comment]
> I'm wondering if we should handle "external shared documents" in the back-end, so that front-ends can still use all the back-end machinery for propagating updates to all users.

Since these custom drives can do very arbitrary actions without requiring a server, it sounds a bit limiting to require a transit through an application server. One of the main reasons for using custom drives is exactly to not require communication with the jupyter server, and be able to make API calls to other services if needed.

Another example of such drive is https://github.com/jupyterlab-contrib/jupyterlab-filesystem-access/

**Comment by davidbrochart** [comment]
I get your point, but I think we should decouple the contents API from the collaborative logic. Right now they are all in the `YDocWebSocketHandler`, but if we had the collaborative logic separate, the front-end could call the corresponding handler (it would not have to, of course).

**Comment by hbcarlos** [comment]
#13907 will partially fix this issue. It will allow using custom drives with RTC enabled but documents coming from a custom drive won't be collaborative.

**Comment by hbcarlos** [comment]
Hi @a3626a! This issue was fixed by #13907.

It is possible to use custom drives and install custom drives next to the [RTC](https://github.com/jupyterlab/jupyter_collaboration) drive. The files opened from the custom drive won't be collaborative, though.

If you want to create a custom drive with real-time collaboration, please, open an issue in the [new RTC repository](https://github.com/jupyterlab/jupyter_collaboration)
